### PR TITLE
Randomisation for loo_pit of discrete data.

### PR DESCRIPTION
**Pull request by TeemuSailynoja**
Fixes issue #22 

I added randomisation to the loo_pit function so that it can be used for discrete data too.

Essentially, when computing pit, we check if the posterior sample contains an exact match to the observation.
In the case an exact match is found, the pit is randomised.

I also added a check that the log weights are finite and not NULL, as well as a warning, if the loo_pit values contain values greater than 1, which can happen with non-normalized weights.

### CODE REVIEW

File: tests/testthat/test-default-methods.R [hunk 1]
```diff
@@ -2,6 +2,13 @@ context("default methods")
 set.seed(1111)
 x <- matrix(rnorm(150), 50, 3)
 y <- rnorm(ncol(x))
+lw <- matrix(rnorm(150), 50, 3)
+lw <- sweep(
+  lw,
+  MARGIN = 2,
+  STATS = matrixStats::colLogSumExps(lw),
```
Comment (jgabry) [inline tests/testthat/test-default-methods.R:9]: We don't have a dependency on matrixStats in rstantools. Can you change this to just do the column log sum exps using base R?

File: R/loo-functions.R [hunk 2]
```diff
@@ -60,11 +60,32 @@ loo_pit.default <- function(object, y, lw, ...) {
...
+  pits <- vapply(seq_len(ncol(yrep)), function(j) {
+    sel_min <- yrep[, j] < y[j]
+    pit <- .exp_log_sum_exp(lw[sel_min, j])
+    sel_sup <- yrep[, j] == y[j]
+    if (any(sel_sup)) {
+      # randomized PIT for discrete y (see, e.g., Czado, C., Gneiting, T.,
+      # Held, L.: Predictive model assessment for count data.
+      # Biometrics 65(4), 1254–1261 (2009).)
+      pit_sup <- pit + .exp_log_sum_exp(lw[sel_sup, j])
+      pit <- stats::runif(1, pit, pit_sup)
```
Comment (jgabry) [inline R/loo-functions.R:75]: Can you mention somewhere in the doc that randomization is used now so the results won't be reproduce without using the same seed?

**Review by jgabry — CHANGES_REQUESTED** [review]
Thanks @TeemuSailynoja! Just a few minor comments.

File: tests/testthat/test-default-methods.R (hunk 1 above)
Comment (jgabry) [reply tests/testthat/test-default-methods.R:9]: I think this should fix the R cmd check issue

**Comment by jgabry** [comment]
Also, can you save the new RDS files using `saveRDS(..., version = 2)`? That should get rid of the message in R cmd check about serialization and adding a dependency on R > 3.5 (https://github.com/stan-dev/rstantools/actions/runs/7798510507/job/21295468864?pr=121#step:6:51). Although maybe we should require R > 3.5 anyway, but we can do that later.

**Comment by TeemuSailynoja** [comment]
Thank you, @jgabry. I hope to have addressed all the requested changes and additions in the last commits.
Also, I fixed an error preventing some of the documentation being parsed, as there was a line break at the `>` in `GPL >= 3`, which caused the line starting with `>` to be parsed as a block quote.

**Review by jgabry — APPROVED** [review]
Thanks @TeemuSailynoja, this looks good!

**Comment by jgabry** [comment]
@andrjohns The StanFunctions Support checks are failing due to a compilation issue with the lgpr package, but that must be unrelated to this PR (this PR doesn't touch anything relevant for that), so I will go ahead and merge and we can sort out whatever is causing those failures before we do a release.

**Pull request merged**